### PR TITLE
Implementation of IgniteableData

### PR DESCRIPTION
**Pull request by hsyyid**
This adds IgniteableData

### CODE REVIEW

**Comment by gabizou** [comment]
Please update according to the changes recently merged.

**Comment by hsyyid** [comment]
@gabizou Give me a bit. :smile:

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+        registerKeyValue(Keys.FIRE_DAMAGE_DELAY, new GetterFunction<ImmutableValue<?>>() {
+
+            @Override
+            public ImmutableValue<?> get() {
+                return fireDelay();
+            }
+        });
+    }
+}
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java]: new line at the end of file.

File: src/main/java/org/spongepowered/common/data/key/KeyRegistry.java [hunk 2]
```diff
@@ -61,6 +62,8 @@ public static void registerKeys() {
         keyMap.put("velocity", makeSingleKey(Vector3d.class, Value.class, of("Velocity")));
         keyMap.put("eye_height", makeSingleKey(Double.class, Value.class, of("EyeHeight")));
         keyMap.put("eye_location", makeSingleKey(Vector3d.class, Value.class, of("EyeLocation")));
+        keyMap.put("fire_ticks", makeSingleKey(Integer.class, Value.class, of("FireTicks")));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/key/KeyRegistry.java]: These need to be MutableBoundedValue, not Value. We need to make sure that we are using the correct types.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/FireDamageDelayValueProcessor.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.data.value.ValueContainer;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.data.value.mutable.MutableBoundedValue;
+import org.spongepowered.common.data.ValueProcessor;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+import org.spongepowered.common.data.value.mutable.SpongeBoundedValue;
+import org.spongepowered.common.data.value.mutable.SpongeValue;
+
+public class FireDamageDelayValueProcessor implements ValueProcessor<Integer, MutableBoundedValue<Integer>> {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/FireDamageDelayValueProcessor.java]: Extend AbstractSpongeValueProcessor.

File: src/main/java/org/spongepowered/common/mixin/core/entity/living/MixinEntityLivingBase.java [hunk 4]
```diff
@@ -105,7 +105,7 @@ public void setMaxHealth(double maxHealth) {
     }
 
     public void damageD(double amount) {
-        Living thisEntity = (Living) this;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/entity/living/MixinEntityLivingBase.java]: Unnecessary change.

File: src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/entity/IgniteableDataBuilder.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public IgniteableData create() {
+        return new SpongeIgniteableData();
+    }
+
+    @Override
+    public Optional<IgniteableData> createFrom(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            final int fireTicks = ((net.minecraft.entity.Entity) dataHolder).fire;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/entity/IgniteableDataBuilder.java]: You should likely surround this logic with `if (dataHolder instanceOf Entity)` for future possible expansions.

File: src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/entity/IgniteableDataBuilder.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            return Optional.<IgniteableData>of(new SpongeIgniteableData(fireTicks, fireDelay));
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public Optional<IgniteableData> build(DataView container) throws InvalidDataException {
+        IgniteableData igniteableData = create();
+        igniteableData.set(Keys.FIRE_DAMAGE_DELAY, getData(container, Keys.FIRE_DAMAGE_DELAY));
+        igniteableData.set(Keys.FIRE_TICKS, getData(container, Keys.FIRE_TICKS));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/entity/IgniteableDataBuilder.java]: Why not just assign the delay and ticks to `int` vars and use them in the `SpongeIgniteableData` constructor?

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        super(ImmutableIgniteableData.class);
+        this.fireTicks = fireTicks;
+        this.fireDelay = fireDelay;
+        registerGetters();
+    }
+
+    @Override
+    public ImmutableBoundedValue<Integer> fireTicks() {
+        return ImmutableDataCachingUtil.getValue(ImmutableSpongeBoundedValue.class, Keys.FIRE_TICKS, this.fireTicks, this.fireTicks, intComparator(),
+                0);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java]: You're missing an extra argument. Put it this way: The `ImmutableDataCachingUtil.getValue()` is supposed to take all the arguments to match a specific constructor of the provided `ImmutableSpongeBoundedValue` class. The parameters must match up exactly, otherwise the util cannot construct the object.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    @Override
+    public ImmutableBoundedValue<Integer> fireDelay() {
+        return ImmutableDataCachingUtil.getValue(ImmutableSpongeBoundedValue.class, Keys.FIRE_DAMAGE_DELAY, this.fireDelay, this.fireDelay,
+                intComparator(), 0);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java]: The same goes for here.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+    @Override
+    public ImmutableBoundedValue<Integer> fireTicks() {
+        return ImmutableDataCachingUtil.getValue(ImmutableSpongeBoundedValue.class, Keys.FIRE_TICKS, this.fireTicks, this.fireTicks, intComparator(),
+                0);
+    }
+
+    @Override
+    public ImmutableBoundedValue<Integer> fireDelay() {
+        return ImmutableDataCachingUtil.getValue(ImmutableSpongeBoundedValue.class, Keys.FIRE_DAMAGE_DELAY, this.fireDelay, this.fireDelay,
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java]: The "default" delay shouldn't be what the actual delay is.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+                intComparator(), 0);
+    }
+
+    @Override
+    public ImmutableIgniteableData copy() {
+        return new ImmutableSpongeIgniteableData(this.fireTicks, this.fireDelay);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java]: return `this`. No need to make more copies.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    public ImmutableSpongeIgniteableData(int fireTicks, int fireDelay) {
+        super(ImmutableIgniteableData.class);
+        this.fireTicks = fireTicks;
+        this.fireDelay = fireDelay;
+        registerGetters();
+    }
+
+    @Override
+    public ImmutableBoundedValue<Integer> fireTicks() {
+        return ImmutableDataCachingUtil.getValue(ImmutableSpongeBoundedValue.class, Keys.FIRE_TICKS, this.fireTicks, this.fireTicks, intComparator(),
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java]: I'm not 100% sure we want to be using the cache util for these types of values. Considering that these values can have many ranges, it doesn't make sense to cache them rather than just creating new ones.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof net.minecraft.entity.Entity;
+    }
+
+    @SuppressWarnings("unused")
+    @Override
+    public Optional<IgniteableData> from(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            final SpongeIgniteableData igniteableData = new SpongeIgniteableData();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: Surround this with an if check for `instanceof Entity`.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+            return Optional.absent();
+        }
+
+    }
+
+    @Override
+    public Optional<IgniteableData> fill(DataHolder dataHolder, IgniteableData manipulator, MergeFunction overlap) {
+        if (supports(dataHolder)) {
+            final IgniteableData merged = overlap.merge(checkNotNull(manipulator).copy(), from(dataHolder).get());
+            manipulator.set(Keys.FIRE_DAMAGE_DELAY, merged.fireDelay().get()).set(Keys.FIRE_TICKS, merged.fireTicks().get());
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: put the last `.set` on a new line to improve readability.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    @Override
+    public Optional<IgniteableData> fill(DataContainer container, IgniteableData igniteableData) {
+        igniteableData.set(Keys.FIRE_TICKS, getData(container, Keys.FIRE_TICKS));
+        igniteableData.set(Keys.FIRE_DAMAGE_DELAY, getData(container, Keys.FIRE_DAMAGE_DELAY));
+        return Optional.of(igniteableData);
+    }
+
+    @Override
+    public DataTransactionResult set(DataHolder dataHolder, IgniteableData manipulator, MergeFunction function) {
+        final int oldFireDamageDelay = manipulator.fireDelay().get();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: A lot of this is not really necessary if the `DataHolder` isn't supported.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        manipulator.set(Keys.FIRE_TICKS, merged.fireTicks().get());
+        final int fireDamageDelay = merged.fireDelay().get();
+        final int fireTicks = merged.fireTicks().get();
+        if (this.supports(dataHolder)) {
+            net.minecraft.entity.Entity entity = (net.minecraft.entity.Entity) dataHolder;
+            entity.fire = fireTicks;
+            entity.fireResistance = fireDamageDelay;
+            Collection<ImmutableValue<?>> successful = new ArrayList<ImmutableValue<?>>();
+            Collection<ImmutableValue<?>> replaced = new ArrayList<ImmutableValue<?>>();
+            successful.add(new ImmutableSpongeValue<Integer>(Keys.FIRE_TICKS, oldFireTicks));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: You're using the wrong type of value here. It's supposed to be an `ImmutableSpongeBoundedValue`.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+            replaced.add(new ImmutableSpongeValue<Integer>(Keys.FIRE_DAMAGE_DELAY, fireDamageDelay));
+            return DataTransactionBuilder.successReplaceResult(successful, replaced);
+        }
+        return DataTransactionBuilder.failResult(manipulator.getValues());
+    }
+
+    @Override
+    public Optional<ImmutableIgniteableData> with(Key<? extends BaseValue<?>> key, Object value, ImmutableIgniteableData immutable) {
+        if (key.equals(Keys.FIRE_DAMAGE_DELAY)) {
+            return Optional.<ImmutableIgniteableData>of(ImmutableDataCachingUtil.getManipulator(ImmutableSpongeIgniteableData.class, value));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: This isn't how the caching data works.... You're not providing enough information for the constructor to be called correctly, let alone the fact that this isn't the best type of data to cache.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+        } else if (key.equals(Keys.FIRE_TICKS)) {
+            return Optional.<ImmutableIgniteableData>of(ImmutableDataCachingUtil.getManipulator(ImmutableSpongeIgniteableData.class, value));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: Same here.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public DataTransactionResult remove(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            if (((net.minecraft.entity.Entity) dataHolder).isRiding()) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: is riding?

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+            }
+        } else {
+            return DataTransactionBuilder.builder().result(DataTransactionResult.Type.FAILURE).build();
+        }
+    }
+
+    @Override
+    public Optional<IgniteableData> createFrom(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            final int fireTicks = ((net.minecraft.entity.Entity) dataHolder).fire;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: Surround by an instanceof check.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/FireTicksValueProcessor.java [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Override
+    public DataTransactionResult removeFrom(ValueContainer<?> container) {
+        return DataTransactionBuilder.failNoData();
+    }
+
+    @Override
+    protected MutableBoundedValue<Integer> constructValue(Integer defaultValue) {
+        return new SpongeBoundedValue<Integer>(this.getKey(), defaultValue, intComparator(), defaultValue, defaultValue);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/FireTicksValueProcessor.java]: No.... please please please read up on what the arguments actually mean!

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+    public ImmutableSpongeIgniteableData(int fireTicks, int fireDelay) {
+        super(ImmutableIgniteableData.class);
+        this.fireTicks = fireTicks;
+        this.fireDelay = fireDelay;
+        registerGetters();
+    }
+
+    @Override
+    public ImmutableBoundedValue<Integer> fireTicks() {
+        return ImmutableDataCachingUtil.getValue(ImmutableSpongeBoundedValue.class, Keys.FIRE_TICKS, this.fireTicks, 0);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java]: Don't use the cache for these types of values.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @Override
+    public ImmutableBoundedValue<Integer> fireDelay() {
+        return ImmutableDataCachingUtil.getValue(ImmutableSpongeBoundedValue.class, Keys.FIRE_DAMAGE_DELAY, this.fireDelay, 20);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java]: Don't use the cache for these types of values.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 23]
```diff
@@ -0,0 +1,151 @@
...
+    public Optional<IgniteableData> fill(DataContainer container, IgniteableData igniteableData) {
+        igniteableData.set(Keys.FIRE_TICKS, getData(container, Keys.FIRE_TICKS));
+        igniteableData.set(Keys.FIRE_DAMAGE_DELAY, getData(container, Keys.FIRE_DAMAGE_DELAY));
+        return Optional.of(igniteableData);
+    }
+
+    @Override
+    public DataTransactionResult set(DataHolder dataHolder, IgniteableData manipulator, MergeFunction function) {
+        if (this.supports(dataHolder)) {
+            final int oldFireDamageDelay = manipulator.fireDelay().get();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: wrap all of this in a try catch for exceptions.

**Comment by hsyyid** [comment]
@gabizou What now?

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+    public ImmutableSpongeIgniteableData(int fireTicks, int fireDelay) {
+        super(ImmutableIgniteableData.class);
+        this.fireTicks = fireTicks;
+        this.fireDelay = fireDelay;
+        registerGetters();
+    }
+
+    @Override
+    public ImmutableBoundedValue<Integer> fireTicks() {
+        return new ImmutableSpongeBoundedValue<Integer>(Keys.FIRE_TICKS, this.fireTicks, 0, intComparator(), 0, 999999999);
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java]: This should be `Integer.MAX_VALUE`, not just a really big number.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    @Override
+    public ImmutableBoundedValue<Integer> fireDelay() {
+        return new ImmutableSpongeBoundedValue<Integer>(Keys.FIRE_DAMAGE_DELAY, this.fireDelay, 20, intComparator(), 0, 999999999);
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeIgniteableData.java]: Same here.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeIgniteableData.java [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+        registerGettersAndSetters();
+    }
+
+    public SpongeIgniteableData() {
+        this(20, 20);
+    }
+
+    @Override
+    public MutableBoundedValue<Integer> fireTicks() {
+        return new SpongeBoundedValue<Integer>(Keys.FIRE_TICKS, this.fireTicks, intComparator(), 0, this.fireTicks);
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeIgniteableData.java]: The maximum value should be `Integer.MAX_VALUE`

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeIgniteableData.java [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    @Override
+    public MutableBoundedValue<Integer> fireDelay() {
+        return new SpongeBoundedValue<Integer>(Keys.FIRE_DAMAGE_DELAY, this.fireDelay, intComparator(), 0, this.fireDelay);
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeIgniteableData.java]: Same here

File: src/main/java/org/spongepowered/common/data/processor/value/entity/FireTicksValueProcessor.java [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Override
+    public DataTransactionResult removeFrom(ValueContainer<?> container) {
+        return DataTransactionBuilder.failNoData();
+    }
+
+    @Override
+    protected MutableBoundedValue<Integer> constructValue(Integer defaultValue) {
+        return new SpongeBoundedValue<Integer>(this.getKey(), defaultValue, intComparator(), 0, 999999999, defaultValue);
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/FireTicksValueProcessor.java]: Same here with the maximum value.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/FireDamageDelayValueProcessor.java [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+    @Override
+    public DataTransactionResult offerToStore(ValueContainer<?> container, Integer value) {
+        final ImmutableValue<Integer> proposedValue = new ImmutableSpongeValue<Integer>(Keys.FIRE_DAMAGE_DELAY, value);
+        if (supports(container)) {
+            final ImmutableValue<Integer> newFireDelayData = new ImmutableSpongeValue<Integer>(Keys.FIRE_DAMAGE_DELAY, value);
+            final ImmutableValue<Integer> oldFireDelayValue = getApiValueFromContainer(container).get().asImmutable();
+            try {
+                ((net.minecraft.entity.Entity) (container)).fireResistance = value;
+                return DataTransactionBuilder.successReplaceResult(oldFireDelayValue, newFireDelayData);
+            } catch (Exception e) {
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/FireDamageDelayValueProcessor.java]: Don't catch an exception here- the only thing that can throw one is `DataTransactionBuilder`, which shouldn't happen (and therefore shouldn't be suppressed).

File: src/main/java/org/spongepowered/common/data/processor/value/entity/FireTicksValueProcessor.java [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+    @Override
+    public DataTransactionResult offerToStore(ValueContainer<?> container, Integer value) {
+        final ImmutableValue<Integer> proposedValue = new ImmutableSpongeValue<Integer>(Keys.FIRE_TICKS, value);
+        if (supports(container)) {
+            final ImmutableValue<Integer> newFireTicksData = new ImmutableSpongeValue<Integer>(Keys.FIRE_TICKS, value);
+            final ImmutableValue<Integer> oldFireTicksValue = getApiValueFromContainer(container).get().asImmutable();
+            try {
+                ((net.minecraft.entity.Entity) (container)).fire = value;
+                return DataTransactionBuilder.successReplaceResult(oldFireTicksValue, newFireTicksData);
+            } catch (Exception e) {
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/FireTicksValueProcessor.java]: Same here

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+                entity.fireResistance = fireDamageDelay;
+                Collection<ImmutableValue<?>> successful = new ArrayList<ImmutableValue<?>>();
+                Collection<ImmutableValue<?>> replaced = new ArrayList<ImmutableValue<?>>();
+                successful.add(new ImmutableSpongeBoundedValue<Integer>(Keys.FIRE_TICKS, oldFireTicks, 0, intComparator(), 0, 999999999));
+                replaced.add(new ImmutableSpongeBoundedValue<Integer>(Keys.FIRE_TICKS, fireTicks, 0, intComparator(), 0, 999999999));
+                successful
+                        .add(new ImmutableSpongeBoundedValue<Integer>(Keys.FIRE_DAMAGE_DELAY, oldFireDamageDelay, 20, intComparator(), 0, 999999999));
+                replaced.add(new ImmutableSpongeBoundedValue<Integer>(Keys.FIRE_DAMAGE_DELAY, fireDamageDelay, 20, intComparator(), 0, 999999999));
+                return DataTransactionBuilder.successReplaceResult(successful, replaced);
+            } catch (Exception e) {
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: Same here

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 32]
```diff
@@ -0,0 +1,157 @@
...
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof Entity;
+    }
+
+    @SuppressWarnings("unused")
+    @Override
+    public Optional<IgniteableData> from(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            if (dataHolder instanceof Entity) {
+                final SpongeIgniteableData igniteableData = new SpongeIgniteableData();
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: Use the two argument constructor here.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java [hunk 33]
```diff
@@ -0,0 +1,157 @@
...
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public DataTransactionResult remove(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            return DataTransactionBuilder.failNoData();
+        } else {
+            return DataTransactionBuilder.builder().result(DataTransactionResult.Type.FAILURE).build();
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/IgniteableDataProcessor.java]: This is exactly the same as `DataTransactionBuilder.failNoData()` - use the former for both cases.

**Comment by hsyyid** [comment]
@Aaron1011 @gabizou 

Was testing further, and found that: `data.set(Keys.FIRE_TICKS, 100)` does nothing, even though it should obviously set the player on fire. However, it was doing this before, perfectly fine. However, `data.get(Keys.FIRE_TICKS).get()` works fine.

**Comment by gabizou** [comment]
> Was testing further, and found that: data.set(Keys.FIRE_TICKS, 100) does nothing, even though it should obviously set the player on fire. However, it was doing this before, perfectly fine. However, data.get(Keys.FIRE_TICKS).get() works fine.

Are you saying `data` being an instance of `IgniteableData` or a `Player`? If you're changing the value in an `IgniteableData`, you need to offer that data back to the `Player`.

**Comment by gabizou** [comment]
One big issue: You're not considering if the entity is actually on fire or not. You need to check if the fire ticks is indeed 0, if it is, they're not on fire (ergo, returning `Optional.absent()`).

**Comment by hsyyid** [comment]
@gabizou Okay. However, data is an instance of IgniteableData and I did offer it back to the player. So, either something is wrong with my test or with my code. Will make it so it returns Optional.absent() if fire ticks is 0 or below (iirc it can be -20) later today.

**Comment by hsyyid** [comment]
@gabizou Okay, I fixed that issue (I think) however, I found the problem with my test. It appears that `data.set(Keys.FIRE_TICKS, 100);` works fine. However, the issue lies with when I offer it to the player. `player.offer(data);`. It seems like either nothing is happening, or the incorrect things are happening.

**Comment by gabizou** [comment]
Please update according to the recent changes for `Entity` specific `DataProcessor`s.

**Comment by hsyyid** [comment]
Okay

**Comment by hsyyid** [comment]
@gabizou Done

File: src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/entity/IgniteableDataBuilder.java [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Override
+    public IgniteableData create() {
+        return new SpongeIgniteableData();
+    }
+
+    @Override
+    public Optional<IgniteableData> createFrom(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            if (dataHolder instanceof Entity) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/entity/IgniteableDataBuilder.java]: Either do a `supports` or do an `instanceof`, you don't need to do both.

File: src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/entity/IgniteableDataBuilder.java [hunk 35]
```diff
@@ -0,0 +1,70 @@
...
+                final int fireTicks = ((Entity) dataHolder).fire;
+                final int fireDelay = ((Entity) dataHolder).fireResistance;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/entity/IgniteableDataBuilder.java]: You're not checking if the entity is actually on fire or not (`fireTicks > 0`)
